django_cloud_tasks/models.py

```diff
@@ -0,0 +1,118 @@
+from django_cloud_tasks import tasks
+
+class Pipeline(models.Model):
+    name = models.CharField(max_length=50)
+
+    def start(self):
+        routines = self.routines.filter(required_routine_vertices__id__isnull=True)
+        for routine in routines:
+            routine.enqueue()
+
+    def revert(self):
+        routines = self.routines.filter(next_routine_vertices__id__isnull=True)
+        for routine in routines:
+            routine.revert()
+
+class Routine(models.Model):
+    class Statuses(models.TextChoices):
+        PENDING = ("pending", "Pending")
+        SCHEDULED = ("scheduled", "scheduled")
+        RUNNING = ("running", "running")
+        COMPLETED = ("completed", "Completed")
+        FAILED = ("failed", "Failed")
+        REVERTING = ("reverting", "Reverting")
+        REVERTED = ("reverted", "Reverted")
+        ABORTED = ("aborted", "Aborted")
+
+    task_name = models.CharField(max_length=50)
+    pipeline = models.ForeignKey(
+        to="django_cloud_tasks.Pipeline",
+        related_name="routines",
+        on_delete=models.PROTECT,
+    )
+    body = models.JSONField(default=dict)
+    attempt_count = models.PositiveIntegerField(default=0)
+    retries_exhausted_limit = models.PositiveIntegerField(null=True)
+    output = models.JSONField(null=True, blank=True)
+    starts_at = models.DateTimeField(null=True, blank=True)
+    ends_at = models.DateTimeField(null=True, blank=True)
+    status = models.CharField(
+        max_length=20,
+        choices=Statuses.choices,
+        default=Statuses.PENDING,
+    )
+    created_at = models.DateTimeField(
+        auto_now_add=True,
+    )
+    updated_at = models.DateTimeField(
+        auto_now=True,
+    )
+    next_routines = models.ManyToManyField(
+        to='Routine',
+        through='RoutineVertice',
+        through_fields=('routine', 'next_routine'),
+        related_name="dependencies"
+    )
+
+    def fail(self, output: str):
+        self.output = output
+        self.status = self.Statuses.FAILED
+        self.ends_at = datetime.now()
+        self.save()
+
+    def complete(self, output: str):
+        self.output = output
+        self.status = self.Statuses.COMPLETED
+        self.ends_at = datetime.now()
+        self.save()
+
+    def enqueue(self):
+        with transaction.atomic():
+            self.status = self.Statuses.SCHEDULED
+            self.starts_at = datetime.now()
+            self.save()
+            tasks.RoutineTask().delay(routine_id=self.pk)
+
+    def revert(self):
+        with transaction.atomic():
+            if self.status == self.Statuses.COMPLETED:
+                self.status = self.Statuses.REVERTING
+                self.save()
+                meta = {"routine_id": self.pk}
+                self.get_task().revert(data=self.output, _meta=meta)
+                return
+
+            self.status = self.Statuses.ABORTED
+            self.save()
+
+    @property
+    def get_task(self) -> Optional[tasks.Task]:
```

Comment:
Since it's a property, there's no need for get_, just `def task` is seamless enough